contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,6 +921,227 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
+		{
+			special_func = &tsql_special_function_list[i];
+			break;
+		}
+	}
+
+	/* If function is not a special function no additional handling required */
+	if (special_func == NULL)
+		return false;
+
+	sys_varcharoid = get_sys_varcharoid();
+	/* Report error in case of invalid argument datatype */
+	for (int i = 0; i < special_func->nargs; i++)
+	{
+		/* for unknown literals consider its type as sys.VARCHAR */
+		inputTypId = (input_typeids[i] == UNKNOWNOID) ? sys_varcharoid : input_typeids[i];
+
+		/* for UDT use its base type for input argument datatype validation */
+		baseTypeId = get_immediate_base_type_of_UDT_internal(inputTypId);
+		if(OidIsValid(baseTypeId))
+		{
+			inputTypId = baseTypeId;
+		}
+
+		type_match = false;
+		for (int j = 0; j < special_func->valid_arg_types[i].len; j++)
+		{
+			if (!OidIsValid(special_func->valid_arg_types[i].valid_types_oid[j]))
+			{
+				/* if common_utility_plugin_ptr is not initialised */
+				if (common_utility_plugin_ptr == NULL)
+					ereport(ERROR,
+							(errcode(ERRCODE_INTERNAL_ERROR),
+							errmsg("Failed to find common utility plugin.")));
+
+				special_func->valid_arg_types[i].valid_types_oid[j] = (*common_utility_plugin_ptr->get_tsql_datatype_oid)(special_func->valid_arg_types[i].valid_types[j]);
+			}
+
+			validTypId = special_func->valid_arg_types[i].valid_types_oid[j];
+
+			if (inputTypId == validTypId)
+			{
+				type_match = true;
+				break;
+			}
+		}
+		if (!type_match)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+				 		 format_type_be(inputTypId), i+1, special_func->formatted_funcname)));
+		}
+	}
+
+	return true;
+}
+
+/*
+ * tsql_func_select_candidate_for_special_func()
+ *
+ * For functions present in special function list, and try to find best candidate 
+ * based on matching return type. Also throw error in case of invalid argument data type.
+ */
+static FuncCandidateList
+tsql_func_select_candidate_for_special_func(List *names, int nargs, Oid *input_typeids, FuncCandidateList candidates)
+{
+	FuncCandidateList			current_candidate, best_candidate;
+	Oid 						expr_result_type;
+	char					   *proc_nsname;
+	char					   *proc_name;
+	bool						is_func_validated;
+	int							ncandidates;
+	Oid							retType;
+
+	DeconstructQualifiedName(names, &proc_nsname, &proc_name);
+
+	is_func_validated = validate_special_function(proc_nsname, proc_name, nargs, input_typeids);
+
+	/* Return NULL if function is not a special function */
+	if (!is_func_validated)
+		return NULL;
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* function based logic to decide return type */
+	expr_result_type = InvalidOid;
+	if (strlen(proc_name) == 4 && strncmp(proc_name,"trim", 4) == 0)
```

Comment:
Hmm, this means, so far only TRIM function needs this handling?

Reply:
Yes, but other string functions such as replace, translate, substring, stuff etc will also going to require this handling.